### DISCUSSION TITLE
"not" keyword prefix for "using" to exclude a specific namespace

### DISCUSSION THREAD

**Discussion by frankhaugen** [dotnet/csharplang#5285] [Language Ideas]
Background
I have often found myself working on a project that has a lot of "duplicate" named types, in closely named namespaces

example (very simplified):
```C#
using MyApplication.Models.Customers;
using MyApplication.Models.Suppliers;
using MyApplication.Models.Contacts;

using AutoMapper;

namespace MyApplication.Mapping;

public class ContactAddressProfile : Profile
{
    public ContactAddressProfile()
    {
        // Address is a type found in both .Customers and .Suppliers
        // ContactInfo is found in .Contacts
        CreateMap<Address, ContactInfo>() ... ; // This will of course cause the application to be un-buildable
    }
}
```
This build issue is remedied by the `using Address = MyApplication.Models.Customers.Address;` syntax, and some do this with almost all models they use in something like a mapping profile, to be able to have the name be clearified. Another solution is to use the entire, (or rather parts of the), namespace. In "Normal" circumstances this is fine, but with the new `global using`, maybe the language shouldn't support exclusion of a namespace. This would also be a way for tools like resharper, to not auto-add the wrong namespace, (yes this have happened with interesting results from time-to-time).

The suggestion
To have a `not using ...` that will exclude itself and it's sub-namespaces from being able to be referenced unless it's fully typed out

**Comment by huoyaoyuan** [comment]
Generally, duplicating names in different namespaces is discouraged, and you can only find very few examples of them. If there are only a few cases, `using Path = System.IO.Path;` is quite enough.

**Reply by The-Futurist** [reply]
I don't see how the very thing namespaces are designed to do can be discouraged. The whole point of namespaces is to reduce the possibility of name collisions.

**Comment by RikkiGibson** [comment]
If you need to access loads of members with duplicated names in another namespace, consider creating an alias for the namespace.

```cs
using C = MyApplication.Models.Customers;

void M(C.Address address) { }
```

**Reply by bernd5** [reply]
Nice way

**Reply by jnm2** [reply]
"loads of" is a colloquialism that is a stronger form of "lots of"
